js/common/bootstrap/modules/buffer.js

```diff
@@ -162,109 +170,99 @@ global.DEFINE_MODULE('buffer', (function () {
-      }
-
-      this.length = +encoding > 0 ? Math.ceil(encoding) : 0;
-      this.parent = subject.parent ? subject.parent : subject;
-      this.offset = offset;
-    } else {
-      // Find the length
-      switch (type = typeof subject) {
-        case 'number':
-          this.length = +subject > 0 ? Math.ceil(subject) : 0;
-          break;
-
-        case 'string':
-          this.length = Buffer.byteLength(subject, encoding);
-          break;
-
-        case 'object': // Assume object is array-ish
-          this.length = +subject.length > 0 ? Math.ceil(subject.length) : 0;
-          break;
-
-        default:
-          throw new TypeError('First argument needs to be a number, ' +
-            'array or string.');
+  const BUFFER_PROXY_HANDLER = {
+    get (target, name) {
+      if (name === '__buffer__') { return true; }
+      if (typeof name === 'string' && name.match(/^\d+$/)) {
+        return target.parent[name];
       }
-
-      // note: the following condition means we'll always create a SlowBuffer
-      // for non-empty lengths. This is required to circumvent using the shared
-      // pool, which is only allocated but never freed (and therefore leaks
-      // memory on shutdown)
-      if (this.length > 0) {
-        // Big buffer, just alloc one.
-        this.parent = new SlowBuffer(this.length);
-        this.offset = 0;
+      return target[name];
+    },
+    set (target, name, value) {
+      if (typeof name === 'string' && name.match(/^\d+$/)) {
+        target.parent[name] = value;
       } else {
-        // Zero-length buffer
-        this.parent = new SlowBuffer(0);
-        this.offset = 0;
+        target[name] = value;
       }
+      return true;
+    }
+  };
 
-      // optimize by branching logic for new allocations
-      if (typeof subject !== 'number') {
-        if (type === 'string') {
-          // We are a string
-          this.length = this.write(subject, 0, encoding);
-        // if subject is buffer then use built-in copy method
-        } else if (Buffer.isBuffer(subject)) {
-          if (subject.parent)
-            subject.parent.copy(this.parent,
-              this.offset,
-              subject.offset,
-              this.length + subject.offset);
-          else
-            subject.copy(this.parent, this.offset, 0, this.length);
-        } else if (isArrayIsh(subject)) {
-          for (var i = 0; i < this.length; i++)
-            this.parent[i + this.offset] = subject[i];
-        }
-      }
+  function createBufferFromString(subject, encoding) {
+    const buffer = Object.create(Buffer.prototype);
+    const length = Buffer.byteLength(subject, encoding);
+    buffer.parent = new SlowBuffer(length);
+    buffer.offset = 0;
+    buffer.length = buffer.parent.write(subject, encoding);
+    return new Proxy(buffer, BUFFER_PROXY_HANDLER);
+  }
+
+  function createBufferFromArrayBuffer(subject, byteOffset = 0, length = subject.byteLength - byteOffset) {
```

Comment:
We previously didn't support ArrayBuffer. Node supports them explicitly so we want that for interop as well.